linalg/solvers.cpp

```diff
@@ -103,6 +104,49 @@ void IterativeSolver::SetOperator(const Operator &op)
 }
 
 
+OperatorJacobiSmoother::OperatorJacobiSmoother(const Vector &d,
+                                               const Array<int> &ess_tdofs,
+                                               const double dmpng)
+   :
+   Solver(d.Size()),
+   N(d.Size()),
+   dinv(N),
+   diag(d),
+   damping(dmpng),
+   ess_tdof_list(ess_tdofs),
+   residual(N) { Setup(); }
+
+void OperatorJacobiSmoother::Setup()
+{
+   residual.UseDevice(true);
+   const double delta = damping;
+   auto D = diag.Read();
+   auto DI = dinv.Write();
+   MFEM_FORALL(i, N, DI[i] = delta / D[i]; );
```

Comment:
This line creates a compilation error. `linalg/solvers.cpp(125): error: The enclosing parent function ("Setup") for an extended __device__ lambda cannot have private or protected access within its class`

Reply:
Does that mean that `Setup()` needs to be public?

Reply:
right

Reply:
Just discussed with @camierjs and apparently this is a limitation of Cuda lambda functions, so `Setup()` has to be public, sorry for suggesting that change.